### ISSUE TITLE
SublimeLinter: syntax_map has been Deprecated

### ISSUE THREAD

**Issue by zwtderek1986** [viluon/LuaExtended#17]
After installing and setting LuaExtended, SublimeLinter can not work properly and lose detection capability.

**Comment by InternetUnexplorer** [comment]
Hmm, that's odd. LuaExtended should be compatible with SublimeLinter.
Did you follow the [instructions](https://github.com/viluon/LuaExtended#luaextended-and-linters) in the README for configuring it?

**Comment by viluon** [comment]
As @InternetUnexplorer said, LuaExtended should be fully compatible with SublimeLinter (there are literally no opportunities for clashes between the two packages, as far as we know).

To be able to fix the issue, we first need to find out what exactly is going on. Would you mind sharing a bit more information? Notably what SublimeLinter plugins are you using, which files did you try to open (does linting fail with Lua files only or does LuaExtended also break the linters for other languages?), whether Sublime Text crashes or shows any errors in dialogues or in the console, and whether you've followed the linked [instructions](https://github.com/viluon/LuaExtended#luaextended-and-linters) for linters.

My best guess is you meant to say that Lua files aren't linted after installing LuaExtended - note that our package takes precedence over the stock Lua syntax (so `*.lua` files open with LuaExtended by default) and needs [a syntax map entry](https://github.com/viluon/LuaExtended#luaextended-and-linters) so that SublimeLinter knows how to treat it. Without this manual change, SublimeLinter doesn't know which linters to apply when LuaExtended is in use.

**Comment by zwtderek1986** [comment]
![1](https://user-images.githubusercontent.com/39580899/40594965-aab7ff4c-6264-11e8-8fa2-781b2d8a915b.png)
![2](https://user-images.githubusercontent.com/39580899/40594967-ab18d88a-6264-11e8-9898-885657eeca4a.png)
![3](https://user-images.githubusercontent.com/39580899/40595032-f251db16-6264-11e8-8775-8f40ec22eba1.png)
SublimeLinter has been updated recently. After setting up, SublimeLinter has reported an error

**Comment by viluon** [comment]
@zwtderek1986 thanks for the clarification. From the error message, it appears that LuaExtended is not the issue here. SublimeLinter-luacheck can't seem to find the luacheck executable. Please make sure that luacheck is in your path and the linter has access to it. If you still think LuaExtended is the cause, try uninstalling it and opening a Lua source file again. I suspect the error will persist.

I'll verify this myself once I find a bit of spare time, I have an exam on linear algebra tomorrow.

**Comment by zwtderek1986** [comment]
@viluon After uninstalling the LuaExtended, Sublimelluacheck runs normally. I tested the problem before committing the bug. I hope this problem can be solved as soon as possible. LuaExtended is a great plug-in! Besides, I hope you take a good exam!

![1](https://user-images.githubusercontent.com/39580899/40632190-d92756b8-6316-11e8-8a8f-58dc41049a6a.png)

The error was that the sublimelinter-luacheck was not set. I deleted the sublimelinter-luacheck. The sublimelinter-lua does fail after installing the LuaExtended, and the sublimelinter-lua works fine after uninstalling the LuaExtended!

**Comment by viluon** [comment]
Thank you @zwtderek1986! Now then, this is very peculiar. I'm starting my own investigation into what is causing this right now. I will update on my progress here once I find out more.

**EDIT**: Also, pardon me for being so dismissive before. I did not expect this to be a major bug.

**Comment by viluon** [comment]
- [x] bug confirmed
- [x] cause identified
- [x] workaround found
- [ ] waiting for maintainer (@viluon)
- [ ] issue resolved
- [ ] patch released

### Bug Confirmed
Happens with SublimeLinter-lua but not with SublimeLinter-luacheck.

### Cause Identified
SublimeLinter-luacheck has [native support for LuaExtended](https://github.com/SublimeLinter/SublimeLinter-luacheck/pull/19) -- something I was not aware of! Very cool :heart_eyes: This is why the bug may not be apparent to all users.

The cause of the issue seems to be a change in SublimeLinter 4.0, which moved from using names for syntax definitions to scope selectors. While this change is wonderful, SL plugins now seem to disregard the `syntax_map` setting.

From the new default settings file for SublimeLinter:
```js
// DEPRECATED: use the selector linter setting instead:
// http://www.sublimelinter.com/en/stable/linter_settings.html#selector
// ...
"syntax_map": {
	...
},
```

### Workaround Found
Scope selectors can be [overridden per-plugin](http://www.sublimelinter.com/en/stable/linter_settings.html#selector). This is a temporary fix to enable LuaExtended linting everywhere.

#### Temporary Solution
Add a "linters" section to your SublimeLinter settings, if it doesn't have one already. Then configure the linters of your choice like so:
```js
"linters": {
	// for SublimeLinter-lua
	"lua": {
		"selector": "source.lua, source.luae"
	},
	// for SublimeLinter-someotherlinter
	"someotherlinter": {
		"selector": "source.lua, source.luae"
	},
},
```

### Resolution is ~~Underway~~ Stalling
I have opened a pull request for the most popular Lua linter out there (https://github.com/SublimeLinter/SublimeLinter-lua/pull/8), and opened an issue to see whether a universally applicable solution is a possibility (https://github.com/SublimeLinter/SublimeLinter/issues/1441).

---

#### Update (4th July)
From the comments I got on the PR, I found out what needs to be done (make LuaExtended replace the official Lua syntax definition and/or merge the two). I'm unfortunately really busy with work at the moment and I can't do that now. I'm also on a vacation next week, hopefully I'll be able to get back to this issue and fix it all when I come back (week 16th - 22nd).

---

### More Changes
Since SublimeLinter-luacheck supports us out of the box, it wouldn't hurt at all to give them a shoutout in the readme! It is also the best Lua linter I have come across thus far.

**Comment by zwtderek1986** [comment]
😍Great. I'm looking forward to solving this problem. Also hope LuaExtended gets better!

**Comment by zwtderek1986** [comment]
In addition, is there any code prompt for "table"? Examples: "table.remove", "table.sort", "table.insert", "table.concat"
![1](https://user-images.githubusercontent.com/39580899/40692562-2625c96c-63e5-11e8-8faa-37f8cdb17a22.png)

**Comment by viluon** [comment]
@zwtderek1986 Yes, [there is](https://github.com/viluon/LuaExtended/blob/master/stdlib.sublime-completions#L74-L78)! I don't know why they aren't showing up for you, however :thinking: They work for me, but I can't provide a screenshot due to technical reasons. Would you mind [opening another issue](https://github.com/viluon/LuaExtended/issues/new)?

**Comment by zwtderek1986** [comment]
ok
